### DISCUSSION TITLE
Get Substitutions before executing the format method

### DISCUSSION THREAD

**Discussion by panayiotisp** [axuno/SmartFormat#140] [Q&A] [answered]
Is there a way from the library itself, to get the list of substitutions  (like the names) prior to executing the format?
I am looking to control which exact values to load from my datastore before using the library in a look of multiple records. 

If there is no existing method I will result to regular expression, but it would be nice if the library had the support so it would provide the substitutions without any control or conditional elements.

**Comment by axunonb** [comment, marked as answer]
>  (like the names) prior to executing the format

If you don't mind executing the formatter with an empty data source, you could get all existing placeholders like shown in the code below.

```csharp
class Program
{
    private static readonly List<string> _placeholders = new List<string>();
    
    static void Main(string[] args)
    {
        Smart.Default.Settings.FormatErrorAction = ErrorAction.Ignore;
        Smart.Default.Settings.ParseErrorAction = ErrorAction.Ignore;
        Smart.Default.OnFormattingFailure += DefaultOnOnFormattingFailure;
        
        var result = Smart.Format("{few} {placeholders} {that} {cannot} {be} {found}", new Dictionary<string, string>());
        Console.WriteLine(result);
        foreach (var placeholder in _placeholders)
        {
            Console.WriteLine(placeholder);
        }
    }

    private static void DefaultOnOnFormattingFailure(object? sender, FormattingErrorEventArgs e)
    {
        _placeholders.Add(e.Placeholder);
    }
}
```
Output:
```
{few}
{placeholders}
{that}
{cannot}
{be}
{found}
```

**Reply by panayiotisp** [reply]
Thanks, 

it makes sense. I was actually experimenting with this as well, although I only set the formatting error to ignore.
but at the same time I wasn't sure if there are expressions in the template string, as opposed to just variable substitutions if they will be accounted as placeholders.

At least this is a good start and I can manipulate if necessary the values in the event handler.

I would like to congratulate you and the rest of the team maintaining this library for excellent work that help all of us.
Again a big thanks!

**Reply by axunonb** [reply]
You're welcome
